[Kernel][Perf] fuse QK Norm and RoPE into one cuda kernel for Qwen Model

izhuhaoran opened the pull request:
## Purpose

Inspired by TensorRT-LLM. This PR is a follow PR about #27018 , and fuses QNorm, KNorm, and RoPE into a single CUDA kernel for the Qwen3 model, improving inference performance. We convert this fusion into a custom torch.compile pass, users can enable it by:

```
 --compilation-config='{"use_inductor": 1,  "pass_config": {"enable_qk_norm_rope_fusion": 1}}'
```

More details see #27018 


### Result GPU Trace

- Main - No Inductor, all custom op
<img width="1317" height="232" alt="image" src="https://github.com/user-attachments/assets/44f636f9-b2ba-4f1e-8fbb-0d68c559d848" />

- Main - use inductor

<img width="1170" height="237" alt="image" src="https://github.com/user-attachments/assets/75cdb5b5-ad5d-4ac8-babe-8abf6d85be25" />


- This PR

<img width="1211" height="233" alt="image" src="https://github.com/user-attachments/assets/f4a679c3-831f-4ed2-be7c-c8a436bba1a5" />

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

izhuhaoran committed:
feat: fuse QNorm KNorm and RoPE into a single cuda kernel

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
refactor: use const for q k weight in fused_qk_norm_rope

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
refactor: update the note comment for fusedQKNormRopeKernel

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
refactor: fix lint error

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
feat: add QKNormRoPEFusionPass for torch.compile

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
add test for qknrom rope fuse pass

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
Merge branch 'main' into fuse-qknorm-rope

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
update qknorm_rope fusion pass and its unit test

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
lint: fix lint error

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran commented on vllm/envs.py:
```diff
@@ -1295,6 +1314,10 @@ def get_vllm_port() -> int | None:
+    # If set, use the fuse QKNorm and RoPE kernel
+    "VLLM_FUSE_QKNORM_AND_ROPE": lambda: bool(
+        int(os.getenv("VLLM_FUSE_QKNORM_AND_ROPE", "0"))
+    ),
```
> We should use pass config instead of env vars.

The PR is still WIP. This environment variable comes from #27018 . This PR is currently attempting to convert it into a custom pass. Users will need to use (as you mentioned):
`--compilation_config='{"use_inductor": 1, "pass_config": {"enable_qk_norm_rope_fusion": 1}}'`
Once this PR is completed, these old contents will be cleaned up.

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,278 @@
+            )
+            q = operator.getitem(split_tuple, 0)
+            k = operator.getitem(split_tuple, 1)
+            v = operator.getitem(split_tuple, 2)
```
I think that this should work, pattern tracing is very close to forward code tracing:
```suggestion
            # split qkv -> q,k,v
            q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
```

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,278 @@
+        ]
+
+        # # Register variants across rope ops and with/without contiguous()
+        # # Ensure view ops are canonicalized to reshape in the traced pattern
```
Was this not needed?

izhuhaoran replied:
removed, done.

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,278 @@
+    @VllmInductorPass.time_and_log
+    def __call__(self, graph: fx.Graph) -> None:
+        if not current_platform.is_cuda_alike():
+            return
```
Silent disablement is not good. Should not be enabled at all on non-cuda-alike platforms

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,278 @@
+                "QK Norm+RoPE fusion enabled, but no Attention layers were discovered."
+            )
+            return
+        layer_name, layer = next(iter(attn_layers.items()))
```
This will only register the pattern using one layer, is that intended? Are we sure this will always pick the same layer also?

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,278 @@
+        rope_op: torch._ops.OpOverload,
+        is_neox: bool,
+    ) -> None:
+        self.layer = layer
```
Are you just using layer for these sizes? If yes, don't save the layer object on the pattern object, just extract the size properties

ProExpertProg commented on vllm/config/compilation.py:
```diff
@@ -546,6 +548,10 @@ def __post_init__(self) -> None:
 
+        if self.pass_config.enable_qk_norm_rope_fusion:
+            self.custom_ops.append("+rms_norm")
+            self.custom_ops.append("+rotary_embedding")
```
Let's try to remove this requirement; definitely for rms_norm and hopefully for RoPE as well although RoPE is less important - I assume all custom RoPE ops would be fused away anyway right?

izhuhaoran replied:
Linked issue: https://github.com/vllm-project/vllm/issues/28042

ProExpertProg reviewed (commented):
This looks like the right approach! Once you're done, please clean up the code and add E2E performance and lm-eval numbers.

izhuhaoran committed:
fix: fix QKNormRoPEFusionPass pattern match

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
Merge remote-tracking branch 'origin/main' into fuse-qknorm-rope-compile

izhuhaoran committed:
fix: attemp to support rope native forward match

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
refactor: clean code

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
lint: fix lint error

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,278 @@
+            q_out = EMPTY_LIKE_OP(q_by_head)
+            q_by_head_contiguous = CONTIGUOUS_OP(q_by_head)
+
+            qn = auto_functionalized(
```
done

izhuhaoran commented on vllm/model_executor/models/qwen3.py:
```diff
@@ -117,6 +119,13 @@ def __init__(
             rope_scaling=rope_scaling,
             dual_chunk_attention_config=dual_chunk_attention_config,
         )
+        # Determine if we can use fused QK norm + RoPE
```
Already cleaned up.

izhuhaoran commented on vllm/model_executor/models/qwen3_moe.py:
```diff
@@ -275,6 +277,13 @@ def __init__(
             rope_scaling=rope_scaling,
             dual_chunk_attention_config=dual_chunk_attention_config,
         )
+        # Determine if we can use fused QK norm + RoPE
```
Already cleaned up. Thanks again

izhuhaoran committed:
fix: adjust TODO comment in rms_norm

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
fix: add dtype check for QKNormRoPEFusionPass and use warning_once

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,244 @@
+            QkNormRopePattern.wrap_trace_fn(
+                pm.fwd_only,
+                QkNormRopePattern.fx_view_to_reshape,
+            ),
```
`fx_view_to_reshape` is a common technique used by Inductor itself to convert view operations to reshape operations for better pattern matching capability. Our implementation here follows the same approach as used in `vllm/compilation/fusion_attn.py`. Since this is a common practice in the compilation pass, I think we don't need to add a specific comment here.

izhuhaoran replied:
ok, already add a note comment for it.

izhuhaoran committed:
fix: update pytest for qk_norm_rope_fusion

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
fix: fix non-nexo style precision issue

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
fix: add is_neox false in pytest for qknorm rope fusion

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
ci: add test_qk_norm_rope_fusion in ci test-pipeline

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
feat: support float16 for qkv and support float32 for cos_sin_cache

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran commented:
> I think we should add the test into `test-pipeline.yaml`

Thanks for the suggestion — I’ve already added it.

izhuhaoran committed:
Merge branch 'main' into fuse-qknorm-rope-compile

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

ProExpertProg commented on vllm/compilation/matcher_utils.py:
```diff
@@ -85,10 +162,12 @@ def forward_custom(
     ) -> torch.Tensor:
         result = torch.empty_like(input)
+        # TODO: support non-contiguous input for RMSNorm and remove this
+        input_contiguous = input.contiguous()
```
I think this might be a ROCm-only requirement - can you guard it and also attach a GH issue link?

ProExpertProg commented on vllm/compilation/pass_manager.py:
```diff
@@ -107,8 +108,12 @@ def configure(self, config: VllmConfig):
                 self.passes += [ActivationQuantFusionPass(config)]
 
             if self.pass_config.enable_attn_fusion:
+                logger.debug("Enabling AttnFusionPass")
```
Remove debug

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,247 @@
+            q_weight,
+            k_weight,
+            cos_sin_cache,
+        ]
```
Can you put this into a method called `get_inputs`?

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,247 @@
+
+        if not current_platform.is_cuda_alike():
+            logger.warning_once("QK Norm+RoPE fusion not enabled: unsupported platform")
+            return
```
This is overly defensive, we should never reach this point on non-cuda-alike platforms

ProExpertProg commented on vllm/_custom_ops.py:
```diff
@@ -329,6 +329,7 @@ def rms_norm(
     out: torch.Tensor, input: torch.Tensor, weight: torch.Tensor, epsilon: float
 ) -> None:
     # TODO: Remove this contiguous call when the kernel is updated to support non-contiguous input
+    # If removed, also need to remove contiguous in MatcherRMSNorm
```
Issue link

ProExpertProg commented on vllm/compilation/qk_norm_rope_fusion.py:
```diff
@@ -0,0 +1,247 @@
+
+FUSED_QK_ROPE_OP = torch.ops._C.fused_qk_norm_rope.default
+SPLIT_SIZES_OP = torch.ops.aten.split_with_sizes.default
+RESHAPE_OP = torch.ops.aten.reshape.default
```
Just use the direct methods on tensors

ProExpertProg reviewed (commented):
Can you also add a unit test for the fused kernel?

izhuhaoran committed:
refactor: clean qk_norm_rope_fusion code

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
test: add unit test for fused_qk_norm_rope kernel

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
chore: add linked issue for todo of rotary_embedding

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran commented:
> Can you also add a unit test for the fused kernel?

I've already added a unit test for the fused kernel and addressed your other comments as well. PTAL when you have time.

ProExpertProg commented on vllm/model_executor/layers/rotary_embedding/base.py:
```diff
@@ -101,35 +101,56 @@ def __init__(
 
-    def forward_native(
-        self,
+    @staticmethod
```
Like RMSNorm, if we want to use the forward method for matching it needs to be static so we can trace through it in the matcher object

ProExpertProg commented on csrc/fused_qknorm_rope_kernel.cu:
```diff
@@ -0,0 +1,426 @@
+    } else {                                                                 \
+      TORCH_CHECK(false, "Unsupported cos_sin_cache dtype: ", cache_dtype);  \
+    }                                                                        \
+  }()
```
Use existing float dispatching macro?

ProExpertProg commented on csrc/fused_qknorm_rope_kernel.cu:
```diff
@@ -0,0 +1,426 @@
+
+  auto qkv_dtype = qkv.scalar_type();
+
+  if (qkv_dtype == torch::kHalf) {
```
Use dispatch macro?

izhuhaoran replied:
Done.

ProExpertProg reviewed (approved):
Two nits in the kernel, otherwise LGTM!

izhuhaoran committed:
fix: use vllm dispatch macro and spport rocm hip

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
fix: fix compile error fo sm75 and remove amd support

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
fix: use define to avoid amd compile

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

izhuhaoran committed:
Merge remote-tracking branch 'origin/main' into fuse-qknorm-rope-compile

Signed-off-by: zhuhaoran <zhuhaoran.zhr@alibaba-inc.com>

JartX commented:
@ZJY0516 @izhuhaoran @tjtanaa yes this PR: https://github.com/vllm-project/vllm/pull/28500 solve the problem

ProExpertProg commented:
I think this is broken: #33295
